Docs/BazelForiOSDevelopers.md

```diff
@@ -0,0 +1,370 @@
+
+In Xcode, command line builds are achieved through the command line interface
+`xcodebuild`.  to build the scheme `ios-app` from `MyProject.xcworkspace`,
+`xcodebuild`, would be invoked as:
+
+```
+xcodebuild -workspace MyProject.xcworkspace -scheme ios-app
+```
+
+Bazel exposes a command line which can also do builds. In Bazel, _every_ target
+is in the global WORKSPACE and there is no notion of schemes.
+```
+bazel build ios-app
+```
+
+The Bazel command line has hundreds of options, which can be found in the
+[Bazel
+documentation](https://docs.bazel.build/versions/master/command-line-reference.html).
+For iOS developers, a set of useful flags is available at
+[bazel-ios-users](https://github.com/ios-bazel-users/ios-bazel-users/blob/master/UsefulFlags.md).
+
+In addition to the default bazel options, it's common to create custom
+configuration settings to customize builds, so the possible Bazel command line
+permutations are endless.
+
+For example, the following build file has conditional `copts` on `:app_store`
+
+```
+objc_library(
+   name = "some",
+   srcs = ["some.m"],
+   copts = select({
+      # For app_store we need to build `some.m` with a special copt
+      ":app_store": ["-DAPPSTORE=1"],
+      "//conditions:default": [],
+    })
+)
+config_setting(
+    name = "app_store",
+    values = { "app_store" : "true" }
+)
+```
+
+This is passed in as a define to Bazel
+```
+bazel build ios-app --define app_store=true
+```
+
+For more information about `config_setting`, please see the [Bazel
+documentation](https://docs.bazel.build/versions/master/be/common-definitions.html#configurable-attributes).
+
+### Configuration of libraries and flags via macros
+
+Most users of Bazel implement a higher level system of macros to encapsulate
+defaults of building librarys and simplify configuration management. 
+
+Bazel provides the pythonic programming language Starlark to implement such
+build system logic. Like rules and aspects, macros are defined in `.bzl` files.
+A macro is a convenient way to call a rule, and not recognized by Bazel in the
+same way a rule is.
+
+_Note: The main distinction between a `.bzl` and a `BUILD` file is `BUILD` files are
+used to create targets by calling macros and rules. `.bzl` files define the
+implementation._
+
+To create a wrapper for `objc_library`, create the file `objc_library.bzl`. The
+following macro restricts the customization, and enforces defaults of the
+native `objc_library` rule.
+
+```
+def objc_library(name, srcs=[], hdrs=[], deps=[], data=[]):
+    """
+    An objc_library that turns on pedantic warnings,
+    and enables modules
+    """
+    native.objc_library(
+        name=name,
+        deps=deps,
+        data=data,
+        srcs=srcs,
+        hdrs=hdrs,
+        enable_modules=True,
+        copts=[
+            "-Wpedantic"
+        ]
+    )
+```
+
+By loading our `objc_library` into a `BUILD` file, it will override the native
+`objc_library` rule, which is automatically imported.
```

Comment:
Do we want to promote shadowing native rules?

Reply:
Absolutely, and it's much easier to read. This is common the python rules and other "official" rules

Reply:
Interesting, we prefix most of our rules that wrap native ones for some reason

Reply:
I think it's probably an implementation detail or choice up to the author. For reference:
https://docs.bazel.build/versions/master/be/python.html
https://github.com/bazelbuild/rules_python
If you preferred to have a prefix from a canned rule set, you can define what a symbol is imported as e.g. `load(":objc_library", pins_objc_library="objc_library")`